### ISSUE TITLE
Bug: Quickly selecting checkboxes on iOS doesn't work correctly

### ISSUE THREAD

**Issue by danielddb** [facebook/react#31260]
React version: 18.0.0 and 17.0.2

## Steps To Reproduce

1. Open https://4txdx2.csb.app/ (React 18) in a device running iOS in Safari / Chrome.
2. Quickly tap one checkbox followed by another one.
3. Notice that if you quickly select one checkbox followed by another one, it checks / unchecks the previous checkbox you were on on.

Link to code example:

React 18.0.0
[Sandbox](https://codesandbox.io/p/sandbox/react-18-checkbox-issue-4txdx2), [Preview](https://4txdx2.csb.app/)

React 17.0.2
[Sandbox](https://codesandbox.io/p/sandbox/react-17-checkbox-issue-yqgs6d), [Preview](https://yqgs6d.csb.app/)

React 16.4.0
[Sandbox](https://codesandbox.io/p/sandbox/react-16-checkbox-issue-jzsk96), [Preview](https://jzsk96.csb.app/)

## The current behavior

There appears to be some sort of race condition where tapping on a controlled / uncontrolled `<input type="checkbox">` quickly after tapping another checkbox updates the original checkbox you tapped rather than the checkbox that was just tapped. It appears to be a timing issue - if you wait long enough between taps, the events are fired on the correct elements.

This works in React 16 but not React 17 and 18.

## The expected behavior

The checkbox state should reflect the checked state of the checkbox you tapped, not the previous checkbox you tapped.

**Comment by saul-atomrigs** [comment]
Hello,
I think this occurs due to the async nature of the `setValues`, leading to possible mismatch in state updates. 
Solution? How about these:
1. Add `key={id}` prop to `<div className="checkbox-row">` --> `<div className="checkbox-row" key={id}>`. This ensures each checkbox is identified as unique
2. (optional) Change parameter `values` to other `prevValues`
```ts
  function setSelectedValues(e, value) {
    const isChecked = e.target.checked;

    setValues((prevValues) =>
      isChecked ? [...prevValues, value] : values.filter((v) => v !== value)
    );
  }
```

**Comment by danielddb** [comment]
Thanks for the suggestion @saul-atomrigs. I went ahead and updated the implementation of the controlled checkboxes example based on your feedback. Unfortunately, the issue still persists even with the updated approach.

Moreover, it's worth noting that this bug also occurs with the uncontrolled checkboxes example I provided. This indicates that the problem isn't specific to whether the checkboxes are controlled or uncontrolled; rather, it points to an underlying issue elsewhere.

**Comment by danielddb** [comment]
After spending hours trying to find a solution, adding this CSS fixes it:

```
input[type="checkbox"],
label {
  touch-action: manipulation;
}
```

**Comment by danielddb** [comment]
The fix in my previous comment only works when you zoom in on your mobile device. If you're zoomed out, it does't work.

**Comment by sambehrens** [comment]
> I think this occurs due to the async nature of the setValues, leading to possible mismatch in state updates.
Solution? How about these:

I don't think this could be the issue, because with only two clicks, after the first click I see the state updating other things and then quickly those things that were updated revert back when the next click of a lower checkbox un-checks the first checkbox. So the state changes are getting in in the correct order.

**Comment by sambehrens** [comment]
The weird thing, is that in this [Sandbox](https://codesandbox.io/s/react-16-checkboxes-xzzl96), [Preview](https://xzzl96.csb.app/) React 16 codesandbox on mobile, the issue doesn't exist, but if you zoom in and try quickly clicking the checkboxes in order, only one will be clicked and the others will do nothing. Then if you add the `touch-action: manipulation` change, it fixes it and it works completely, both zoomed out and zoomed in. [Sandbox](https://codesandbox.io/p/sandbox/react-16-checkboxes-forked-ndgrm4), [Preview](https://ndgrm4.csb.app/).

That indicates that it must be an issue with React 18, right?

**Comment by danielddb** [comment]
@sambehrens you're right, I just created the exact same [Sandbox](https://codesandbox.io/p/sandbox/jzsk96), [Preview](https://jzsk96.csb.app/) with React 16 and it works without having to zoom in or adding `touch-action: manipulation`! So it definitely looks like an issue with React 18.

**Comment by sambehrens** [comment]
This seems like a pretty basic flaw in React, so I don't want it to get auto-closed for no activity. How do we get maintainers to acknowledge this issue?

**Comment by sambehrens** [comment]
Just want to make sure that this original issue that I referenced in my issue is linked here too: #27475

**Comment by sambehrens** [comment]
I just tested it on `19.0.0-beta-26f2496093-20240514` react and react-dom and the issue still exists.

**Comment by Ronsku** [comment]
I have the same issue on a regular button to increase or decrease a value (-) (+). If you spam click it will only update once and never again until you take a break and click again.
Feels that there's like a 500ms throttle or something to not cause multiple clicks by mistake, but when that's what you want, what do you do then?

**Comment by danielddb** [comment]
@mizdra see my original comment with the React 16 example, it works fine there. Just not in 17 and 18.

**Comment by mizdra** [comment]
@danielddb The same applies to #19841. #19841 reports that the problem does not reproduce the problem in React 16, but does reproduce it in React 17.0.0-rc.1.

React handles events in a special method (a.k.a. Event Delegation), and the way it manages them has changed slightly in React 17. The Webkit bug does not cause problems with the general event handling method, but it does cause problems with the handling method in React 17 and later.

#19841 has more detailed information. Please read it.

**Comment by LukeLaScala** [comment]
bump. Running into this still.

**Comment by SahilSinghTomar** [comment]
i just ran into this issue.

**Comment by AIC-HMV** [comment]
Pushing is fun, but the core remains untouchable.
You can test, you can patch, but Supreme always stays one move ahead.
Welcome to the playground!
